spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala

```diff
@@ -579,14 +709,21 @@ case class DeltaSharingSource(
     val params = new RemoteDeltaFileIndexParams(
       spark, initSnapshot, deltaLog.client.getProfileProvider)
     val fileIndex = new RemoteDeltaBatchFileIndex(params, addFilesList)
+    val urlExpirationTimestamp = getMinUrlExpirationTimestamp()
```

Comment:
`sortedFetchedFiles`, `lastQueryTableTimestamp` and `minUrlExpirationTimestamp` should probably be accessed together in the same `synchronized`. Right now there is a race condition here: if a refresh is done before line 712, then `urlExpirationTimestamp` will be for the latest URLs, but `idToUrl` is still using the old URLs. This would cause us to use a wrong bigger `urlExpirationTimestamp`. Since `CachedTableManager` sees a larger `urlExpirationTimestamp`, it will not refresh before these old URLs expire and cause token expiration issue.

Reply:
Good catch. Returned the three parameters together from `popSortedFetchedFiles`.